src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -180,18 +165,82 @@ public void start(Promise<Void> future) throws Exception {
                                             event.getEventType().toString());
                             }
                         });
+
+        this.addTargetDiscoveryListener(
+                tde -> {
+                    switch (tde.getEventKind()) {
+                        case MODIFIED:
+                            testNonConnectedTargets(
+                                    entry ->
+                                            Objects.equals(
+                                                    tde.getServiceRef(),
+                                                    entry.getKey().getTarget()));
+                            break;
+                        case LOST:
+                            var it = nonConnectableTargets.entrySet().iterator();
+                            while (it.hasNext()) {
+                                var entry = it.next();
+                                if (Objects.equals(
+                                        tde.getServiceRef(), entry.getKey().getKey().getTarget())) {
+                                    it.remove();
+                                }
+                            }
+                            break;
+                        default:
+                            break;
+                    }
+                });
+    }
+
+    private void checkNonConnectedTargetJvmIds() {
+        testNonConnectedTargets(
+                entry -> {
+                    TargetNode targetNode = entry.getKey();
+                    return testJvmId(targetNode.getTarget());
+                });
+    }
+
+    private boolean testJvmId(ServiceRef serviceRef) {
+        try {
+            String id = jvmIdHelper.get().resolveId(serviceRef).getJvmId();
+            return StringUtils.isNotBlank(id);
+        } catch (JvmIdGetException e) {
+            logger.info("Retain null jvmId for target [{}]", serviceRef.getServiceUri());
+            logger.info(e);
+            return false;
+        }
     }
 
     private void testNonConnectedTargets(Predicate<Entry<TargetNode, UUID>> predicate) {
-        ForkJoinPool.commonPool()
-                .execute(
-                        () -> {
-                            Map<TargetNode, UUID> copy = new HashMap<>(nonConnectableTargets);
-                            for (var entry : copy.entrySet()) {
+        Map<Pair<TargetNode, UUID>, ConnectionAttemptRecord> copy =
+                new HashMap<>(nonConnectableTargets);
+        for (var entry : copy.entrySet()) {
+            ForkJoinPool.commonPool()
```

Comment:
updated to using Executor in your platform detection PR so we might want to update that here too